Use concurrency to check mangled passwords

rafasc opened the pull request:
This is part of an attempt to improve crunchy's performance.

This PR focuses on improving the mangled password checks by using concurrency to speed up the search.
```
$ benchstat before after 
name                   old time/op  new time/op  delta
ValidatePassword-8      41.3s ± 1%   29.7s ± 1%  -28.12%  (p=0.000 n=8+8)
FoundInDictionaries-8   11.8s ± 1%    4.4s ± 1%  -62.24%  (p=0.000 n=8+8)
```


Hashing will be next

muesli commented on crunchy_test.go:
```diff
@@ -13,16 +13,25 @@ import (
 		expected error
 		rating   uint
 	}{
+		// include values from pass in tests
```
What's the purpose of including these values here (again)?

rafasc replied:
Having a reference to a password known to be valid is useful for benchmarks because they never exit earlier and need to go though every check we perform.

We could search the `pws` slice for a valid password when we need one, or make a convention that the first password of the slice is valid. I thought being explicit would make things more clear. I.e. `pass.valid vs pws[0].pw`

The values of the `pass` struct are then inserted into the `pws` slice to make sure they remain valid. If a future change caused `pass.valid` to be invalid, tests would fail, indicating that benchmark numbers are not trustworthy.

muesli commented on crunchy.go:
```diff
@@ -142,12 +143,27 @@ func (v *Validator) foundInDictionaries(s string) error {
+				go func(word string) {
+					var err *DictionaryError
+					if dist := smetrics.WagnerFischer(word, pw, 1, 1, 1); dist <= v.options.MinDist {
+						err = &DictionaryError{ErrMangledDictionary, word, dist}
```
We never check the error returned here. We can already abort the check here if an error occurred. Also, why don't we simply write to `errc` here directly?

rafasc replied:
The intention was to avoid sending more than one value over the channel per run, because that could potentially leak goroutines as the `errc` channel has a buffer of `njobs`.

I think I can make this part less ambiguous with your suggestion. Will push a fix.

muesli commented on crunchy.go:
```diff
@@ -142,12 +143,27 @@ func (v *Validator) foundInDictionaries(s string) error {
-			if dist := smetrics.WagnerFischer(word, revpw, 1, 1, 1); dist <= v.options.MinDist {
-				return &DictionaryError{ErrMangledDictionary, word, dist}
+			select {
+			case queue <- struct{}{}:
```
I'm not sure I fully understand the purpose of this queue. Couldn't we just keep launching a couple of worker threads in a pool and feed them with all the words from the wordlist? I think it may turn out a bit more readable. Let me know if you already tried that but this queue simply outperforms it :smiley:

rafasc replied:
I tried some approaches, e.g. using `x/sync/errgroup`, and this was the simplest way I found to make it concurrency bounded with the ability to cancel on the first error found.

The `queue` is just a semaphore that keeps us from spawning more than `njobs` goroutines simultaneously.

rafasc committed:
Fix benchmark divergence

The benchmark should run with the same input each time in order to
obtain a meaningful average.

Let's also benchmark the cases where hash verification is enabled and
use a valid password as input since they cover more checks.

CheckHIBP is intentionally left out from this benchmark as network
factors could fudge the results.

rafasc committed:
Benchmark dictionary indexing independently

Indexing only happens once per validator. Let's benchmark indexing
independently so we can ignore its duration in other benchmarks.

rafasc committed:
Add benchmark for foundInDictionaries

In a future commit we will improve this function by converting into go
routines.

Introducing the benchmark now is convenient to provide a baseline for
comparison.

rafasc committed:
Use concurrency to check mangled passwords

muesli commented:
Sorry @rafasc, I have never seen any notification that you responded to my comments :disappointed: ...and so I completely missed your latest changes.
